File: day01/main.nim
import future, re, sets, strutils

type Coord = tuple[x: int, y: int]

proc walk(input: string, stopping: (me: Coord) -> bool): int =
  var me: Coord = (x: 0, y: 0)
  var dx, dy, t = 0
  dy = -1

  for m in findAll(input, re"[LR]\d+"):
    case m[0]
      of 'L': t = dx; dx = dy; dy = -t
      else: t = dx; dx = -dy; dy = t
    for i in 1..parseInt(m[1 .. m.high]):
      me.x += dx
      me.y += dy
      if stopping(me):
        return abs me.x + abs me.y

  return abs me.x + abs me.y

var input = readFile "input.txt"
proc neverStop(me: Coord): bool = false
echo "Answer #1: ", walk(input, neverStop)

var places = initSet[Coord]()
proc visited(me: Coord): bool =
  if places.contains(me): return true
  places.incl(me)
  return false
echo "Answer #2: ", walk(input, visited)

File: day01/main.py
import re

def walk(input):
    x, y = 0, 0
    dx, dy = 0, -1
    for m in re.finditer(r'([LR])(\d+)', input):
        if m.group(1) == 'L':
            dx, dy = dy, -dx
        else:
            dx, dy = -dy, dx
        for i in range(0, int(m.group(2))):
            x += dx
            y += dy
            yield (x, y)

for x, y in walk(open('input.txt').read()):
    pass
print("Answer #1:", abs(x) + abs(y))

seen = set()
for x, y in walk(open('input.txt').read()):
    if (x, y) in seen:
        break
    seen.add((x, y))
print("Answer #2:", abs(x) + abs(y))
